File: media-service/media-service/src/main/java/com/inkwell/media/service/MediaService.java
package com.inkwell.media.service;

import com.inkwell.media.dto.MediaDTO;
import com.inkwell.media.dto.UploadResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for media operations.
 * Defines contracts for file upload, retrieval, linking, and soft deletion.
 */
public interface MediaService {

    /**
     * Upload a new media file.
     * Validates file type and size, saves to filesystem, and stores metadata in database.
     *
     * @param file the file to upload
     * @param uploaderId the ID of the uploader
     * @param altText optional alternative text for accessibility
     * @return upload response with file metadata
     */
    UploadResponse uploadMedia(MultipartFile file, Integer uploaderId, String altText);

    /**
     * Retrieve media metadata by ID.
     * Excludes soft-deleted media.
     *
     * @param mediaId the ID of the media
     * @return the media DTO
     */
    MediaDTO getMediaById(Integer mediaId);

    /**
     * Retrieve media by unique filename.
     *
     * @param filename the unique filename
     * @return the media DTO
     */
    MediaDTO getMediaByFilename(String filename);

    /**
     * Get all media files uploaded by a user.
     * Excludes soft-deleted files.
     *
     * @param uploaderId the ID of the uploader
     * @return list of media DTOs
     */
    List<MediaDTO> getMediaByUploaderId(Integer uploaderId);

    /**
     * Get all media files linked to a specific post.
     * Excludes soft-deleted files.
     *
     * @param postId the ID of the post
     * @return list of media DTOs
     */
    List<MediaDTO> getMediaByPostId(Integer postId);

    /**
     * Link a media file to a post.
     *
     * @param mediaId the ID of the media
     * @param postId the ID of the post to link
     */
    void linkToPost(Integer mediaId, Integer postId);

    /**
     * Unlink a media file from any post.
     *
     * @param mediaId the ID of the media
     */
    void unlinkFromPost(Integer mediaId);

    /**
     * Soft delete a media file by marking it as deleted.
     * Preserves the database record for audit purposes.
     * The physical file is also deleted from the filesystem.
     *
     * @param mediaId the ID of the media to delete
     */
    void deleteMedia(Integer mediaId);

    /**
     * Permanently delete a media file (hard delete).
     * Use with caution as this cannot be undone.
     *
     * @param mediaId the ID of the media to permanently delete
     */
    void permanentlyDeleteMedia(Integer mediaId);

    /**
     * Download a media file by its ID.
     * Returns the file content as a byte array.
     *
     * @param mediaId the ID of the media
     * @return the file content as bytes
     */
    byte[] downloadMedia(Integer mediaId);
    /**
     * Get all media files in the system.
     * Excludes soft-deleted files.
     *
     * @return list of all media DTOs
     */
    List<MediaDTO> getAllMedia();
}

File: frontend/src/services/mediaService.ts
import { AxiosProgressEvent } from "axios";
import apiClient from "../api/axiosInstance";
import { Media, UploadResponse } from "../types/media";

const mediaService = {
  upload: async ({
    file,
    uploaderId,
    altText,
    onUploadProgress
  }: {
    file: File;
    uploaderId: number;
    altText?: string;
    onUploadProgress?: (progressEvent: AxiosProgressEvent) => void;
  }) => {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("uploaderId", String(uploaderId));

    if (altText?.trim()) {
      formData.append("altText", altText.trim());
    }

    const { data } = await apiClient.post<UploadResponse>("/api/v1/media/upload", formData, {
      headers: {
        "Content-Type": "multipart/form-data"
      },
      timeout: 15000,
      onUploadProgress
    });

    return data;
  },

  getById: async (mediaId: number) => {
    const { data } = await apiClient.get<Media>(`/api/v1/media/${mediaId}`);
    return data;
  },

  getByPost: async (postId: number) => {
    const { data } = await apiClient.get<Media[]>(`/api/v1/media/post/${postId}`);
    return data;
  },

  getByUploader: async (uploaderId: number) => {
    const { data } = await apiClient.get<Media[]>(`/api/v1/media/uploader/${uploaderId}`);
    return data.filter((media) => !media.isDeleted);
  },

  linkToPost: async (mediaId: number, postId: number) => {
    await apiClient.put(`/api/v1/media/${mediaId}/link/${postId}`);
  },

  unlinkFromPost: async (mediaId: number) => {
    await apiClient.put(`/api/v1/media/${mediaId}/unlink`);
  },

  delete: async (mediaId: number) => {
    await apiClient.delete(`/api/v1/media/${mediaId}`);
  }
};

export default mediaService;
